### DISCUSSION TITLE
Add Member Login/Logout notifications

### DISCUSSION THREAD

**Discussion by Migaroez** [umbraco/Umbraco-CMS#13081] [Features and ideas]
### Origin
In the CodeCabin chat the following question was asked by @MMasey "Does anyone know of any Umbraco packages which track member logins? and include this info in the backoffice?"

After a bit of investigation, it seems like there currently aren't any member login/logout notifications available.

### Possible solution
As far as I can see, the best place to do this would be in the UmbracoSignInManager in HandleSignIn() as indicated by the comment at the top of that function 😄
An other option would be to implement PasswordSignInAsync(), SignInAsync() and SignOutAsync() on MemberSignInManager that wraps its base(UmbracoSignInManager) method with a notification based on the result before passing the result along. But then we should maybe also do this for the BackOfficeSignInManager and we have to write the if/else-if flow again to know which notification to send based on the MC base result...

Maybe we should also include the type of user in the logs, as it was unclear before diving into this code that Member logins are actually logged.

### Possible use cases
- Package that stores the user logins/logouts in storage somewhere and displays this information in a member content app
- Package that sends tracks time between logins for a user and notifies somebody to follow up on them if it has been a long time
- Start a data sync when a particular type of member logs in so the information is ready up to date before they go and visit the particular page
- Show different content to a member if they haven't logged in for a while
- ...

### Resolution
I wouldn't mind implementing this, but would love to have an OK from HQ first and what route they prefer I should take. Any input is greatly appreciated.

**Comment by Bakersbakebread** [comment]
Would be great to add these for Users too, for all the above use cases.

**Comment by bergmania** [comment]
Sounds like a good idea to me.

Implementation wise, the challenge is to make it in a non-breaking way, that is most likely sub-optimal.
Ideally, we would follow a pattern like in `ContentTypeServiceBase<TRepository, TItem>` and implement that in `UmbracoSignInManager` but that would definitely be breaking.

I think the best way to archive it, is to directly make the overrides in `MemberSignInManager`, like we do in `BackOfficeSignInManager` to make the notifications like `UserLoginSuccessNotification`.

I think we can follow the same pattern for logouts, but that is currently not handled for users either.

Please note that we need to handle external logins too.
